server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java

```diff
@@ -49,14 +50,30 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
                 return allocation.decision(Decision.YES, NAME, "shard is currently being restored");
             }
         }
+
+        /**
+         * POST: the RestoreInProgress.ShardRestoreStatus is either failed or succeeded. This section turns a
+         * turn a shard failure into a NO decision to allocate. See {@link AllocationService.applyFailedShards}
+         * for details on how it updates UnassignedInfo.
+         */
+        UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
+        if (unassignedInfo.failedAllocations() > 0) {
+            return allocation.decision(
+                Decision.NO,
+                NAME,
+                "shard has failed to be restored from the snapshot [%s] - manually close or delete the index [%s] in order to retry "
+                    + "to restore the snapshot again or use the reroute API to force the allocation of an empty primary shard. Check the "
+                    + "logs for more information about the failure. Details: [%s]",
+                source.snapshot(),
+                shardRouting.getIndexName(),
+                unassignedInfo.details()
+            );
+        }
+
         return allocation.decision(
-            Decision.NO,
+            Decision.YES,
```

Comment:
This changes the behaviour of the decider and I don't think we should do that. It's important that we return `NO` if the restore UUID isn't in the cluster state any more. We just need to explain that in this case the restore didn't fail, we just never even attempted it before the restore concluded.